pkg/gateway/server/logout_all.go

```diff
@@ -0,0 +1,114 @@
+package server
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/obot-platform/obot/pkg/api"
+	kcontext "github.com/obot-platform/obot/pkg/gateway/context"
+	"github.com/obot-platform/obot/pkg/proxy"
+	v1 "github.com/obot-platform/obot/pkg/storage/apis/obot.obot.ai/v1"
+	"github.com/obot-platform/obot/pkg/system"
+	"github.com/tidwall/gjson"
+)
+
+func (s *Server) logoutAll(apiContext api.Context) error {
+	logger := kcontext.GetLogger(apiContext.Context())
+
+	sessionID, sessionIDFound := getSessionID(apiContext.Request)
+
+	identities, err := s.client.FindIdentitiesForUser(apiContext.Context(), apiContext.UserID())
+	if err != nil {
+		return err
+	}
+
+	var errs []error
+	for _, identity := range identities {
+		if identity.AuthProviderName == "" || identity.AuthProviderNamespace == "" {
+			continue
+		}
+
+		var ref v1.ToolReference
+		if err := apiContext.Get(&ref, identity.AuthProviderName); err != nil {
+			errs = append(errs, fmt.Errorf("failed to get auth provider %q: %w", identity.AuthProviderName, err))
+			continue
+		}
+
+		user := identity.ProviderUserID
+		if identity.AuthProviderName == "github-auth-provider" && identity.AuthProviderNamespace == system.DefaultNamespace {
+			// The GitHub auth provider stores the username as the user ID in the sessions table.
+			// This is because of an annoying quirk of the oauth2-proxy code for GitHub,
+			// where we do not know the real user ID until after the user has logged in and the session is created,
+			// and we have to manually fetch it from the GitHub API.
+			// The oauth2-proxy is only aware of the username, which is why that's in the sessions table.
+			user = identity.ProviderUsername
+		}
+
+		emailHash := sha256.Sum256([]byte(identity.Email))
+		userHash := sha256.Sum256([]byte(user))
+
+		logger.Debug("deleting sessions for provider", "provider", identity.AuthProviderName)
+
+		if meta, ok := ref.Status.Tool.Metadata["providerMeta"]; ok {
+			tablePrefix := gjson.Get(meta, "postgresTablePrefix").String()
+			if tablePrefix != "" {
+				var err error
+				if sessionIDFound {
+					err = s.db.WithContext(apiContext.Context()).Exec("DELETE FROM "+tablePrefix+"sessions WHERE key NOT LIKE ? AND \"user\" = decode(?, 'hex') AND \"email\" = decode(?, 'hex')", sessionID+"%", fmt.Sprintf("%x", userHash), fmt.Sprintf("%x", emailHash)).Error
```

Comment:
Any reason we are not directly comparing the value for user and email in query, instead of doing hash and decode with hash too?

Reply:
yeah, PII can't be stored in plaintext, and email and username are considered PII

Reply:
the decode call in there is just to convert from a hex string representation of a hash, to the postgres `bytea` type